libdnf/rpm/package_sack.cpp

```diff
@@ -75,6 +77,112 @@ void PackageSack::Impl::make_provides_ready() {
     provides_ready = true;
 }
 
+void PackageSack::Impl::setup_excludes_includes(bool only_main) {
+    considered_uptodate = false;
+
+    auto & main_config = base->get_config();
+
+    auto & disable_excludes = main_config.disable_excludes().get_value();
+    std::set<std::string> disabled(disable_excludes.begin(), disable_excludes.end());
+
+    if (disabled.count("all") != 0) {
+        return;
+    }
+
+    PackageSet repo_includes(base);
+    PackageSet repo_excludes(base);
+    bool includes_used = false;   // packages for inclusion specified (they may or may not exist)
+    bool excludes_exist = false;  // found packages for exclude
+    bool includes_exist = false;  // found packages for include
+
+    // first evaluate repo specific includes/excludes
+    if (!only_main) {
+        for (const auto & repo : base->get_repo_sack()->get_data()) {
+            if (!repo->is_enabled() || disabled.count(repo->get_id()) != 0) {
+                continue;
+            }
+
+            if (!repo->get_config().includepkgs().get_value().empty()) {
+                repo->set_use_includes(true);
+                includes_used = true;
+            }
+
+            PackageQuery query_repo_pkgs(base, PackageQuery::Flags::IGNORE_EXCLUDES);
+            query_repo_pkgs.filter_repo_id({repo->get_id()});
+
+            for (const auto & name : repo->get_config().includepkgs().get_value()) {
+                PackageQuery query(query_repo_pkgs);
+                ResolveSpecSettings settings{
+                    .ignore_case = false, .with_nevra = true, .with_provides = false, .with_filenames = false};
+                const auto & [found, nevra] = query.resolve_pkg_spec(name, settings, false);
+                if (found) {
+                    repo_includes |= query;
+                    includes_exist = true;
+                }
+            }
+
+            for (const auto & name : repo->get_config().excludepkgs().get_value()) {
+                PackageQuery query(query_repo_pkgs);
+                ResolveSpecSettings settings{
+                    .ignore_case = false, .with_nevra = true, .with_provides = false, .with_filenames = false};
+                const auto & [found, nevra] = query.resolve_pkg_spec(name, settings, false);
+                if (found) {
+                    repo_excludes |= query;
+                    excludes_exist = true;
+                }
+            }
+        }
+    }
+
+    // then main (global) includes/excludes because they can mask
+    // repo specific settings
+    if (disabled.count("main") == 0) {
+        for (const auto & name : main_config.includepkgs().get_value()) {
+            PackageQuery query(base, PackageQuery::Flags::IGNORE_EXCLUDES);
```

Comment:
The `query` here and in the `for` loop below this one could be moved out of the loops too?

Reply:
The following code `query.resolve_pkg_spec(name, settings, false);` modifies the query. I need a newly initialized query in each cycle.

Reply:
Oh, does it? I assumed it doesn't just because of the name and what I think it does. This is treacherous...